File: src/main/java/net/kigawa/kutil/unit/UnitInfo.java
package net.kigawa.kutil.unit;

import java.lang.reflect.Constructor;

public class UnitInfo
{
    protected final Class<?> unitClass;
    protected Object unit;

    public UnitInfo(Class<?> unitClass) {
        this.unitClass = unitClass;
    }


    protected Constructor<?> getConstructor() {
        var constructors = unitClass.getConstructors();
        if (constructors.length == 1) return constructors[0];
        for (var constructor : constructors) {
            if (constructor.isAnnotationPresent(Inject.class)) {
                return constructor;
            }
        }

        throw new UnitException("cold not get constructor: " + unitClass);
    }
}

File: src/main/java/net/kigawa/kutil/unit/container/UnitInfo.kt
package net.kigawa.kutil.unit.container

import net.kigawa.kutil.unit.factory.UnitFactory
import java.util.concurrent.FutureTask

class UnitInfo(val unitClass: Class<*>, val name: String?) {
    var status: UnitStatus = UnitStatus.NOT_LOADED
        private set

    var unit: Any? = null
        set(value) {
            field = value
            status = UnitStatus.INITIALIZED
        }
    var future: FutureTask<*>? = null
        set(value) {
            field = value
            status = UnitStatus.INITIALIZING
        }
    var factory: UnitFactory? = null
        set(value) {
            field = value
            status = if (field == null) UnitStatus.NOT_LOADED
            else UnitStatus.LOADED
        }

    constructor(unitClass: Class<*>) : this(unitClass, null)

}